Introduce jest test

nonumpa opened the pull request:
1. Test hocuspocus elasticsearch and snapshot extension
    - prepare test DB
    - use hocuspocus [server](https://github.com/ueberdosis/hocuspocus/blob/main/tests/utils/newHocuspocus.ts)/[provider](https://github.com/ueberdosis/hocuspocus/blob/main/tests/utils/newHocuspocusProvider.ts) to the test extension
2. Use github action to test on push and pull request

nonumpa committed:
Import jest

npm install -D jest ts-jest @types/jest

nonumpa committed:
Add jest config

npx ts-jest config:init

nonumpa committed:
Test elasticsearch extension

nonumpa committed:
Add submodule rumors-db

nonumpa committed:
Jest coverage should ignore test folder

nonumpa committed:
Update document about testing

nonumpa committed:
Add test folder to workspaces, so it can be imported into `__test__` script

nonumpa committed:
Add test utils

nonumpa committed:
Fix lint

nonumpa committed:
Lint should ignore submodule

nonumpa committed:
Add github ci yml

nonumpa committed:
Add ci and coverage badge

MrOrz commented on README.md:
```diff
@@ -1,4 +1,7 @@
 # collab-server
+
+[![CI test](https://github.com/cofacts/collab-server/actions/workflows/ci.yml/badge.svg)](https://github.com/cofacts/collab-server/actions/workflows/ci.yml) [![Coverage Status](https://coveralls.io/repos/github/cofacts/collab-server/badge.svg?branch=test)](https://coveralls.io/github/cofacts/collab-server?branch=test)
```
We probably need to remove the `?branch` parameter so that the status points to master branch

```suggestion
[![CI test](https://github.com/cofacts/collab-server/actions/workflows/ci.yml/badge.svg)](https://github.com/cofacts/collab-server/actions/workflows/ci.yml) [![Coverage Status](https://coveralls.io/repos/github/cofacts/collab-server/badge.svg)](https://coveralls.io/github/cofacts/collab-server)
```

MrOrz commented on package.json:
```diff
@@ -13,7 +14,11 @@
     "build:extension-es": "npm run build -w hocuspocus-extension-elasticsearch",
     "build": "npm run build --workspaces --if-present && tsc -b",
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "pretest": "npm run rumors-db:install && npm run rumors-db:test && mkdir -p build",
```
We probably don't need `mkdir -p build` here.
(Actually I forgot why we have it back in rumors-api......)

nonumpa replied:
fixed

MrOrz commented on hocuspocus-extension-elasticsearch/src/__test__/elasticsearch.test.ts:
```diff
@@ -0,0 +1,86 @@
+    await delayForMs(1000);
+  });
+
+  it('logs error', (resolve) => {
```
Question: do we assert anything in this test?

nonumpa replied:
No, it seems that we can't catch extension error in hocuspocus server, so I just log error instead of throwing it. This test is just run the `catch` part to increase coverage.

MrOrz replied:
Got it. Maybe consider `jest.spyOn` console.log / console.error and expect console is called?

Another alternative would be just [adding `/* istanbul ignore next */` on error handlers](https://github.com/cofacts/rumors-line-bot/pull/376/files#diff-3d74613acb1d956e12832d77da2626ce51445b2b2ced75beccd03847b125f878), if the error does not happen that frequently.

MrOrz commented on hocuspocus-extension-elasticsearch/src/__test__/elasticsearch.test.ts:
```diff
@@ -0,0 +1,86 @@
+  it('return default ydoc when fetched documentName does not exist', (resolve) => {
+    const provider = newHocuspocusProvider(server, {
+      onSynced() {
+        expect(provider.document.share.size).toBe(0);
```
Question: what is `document.share.size` here?

nonumpa replied:
[`document.share`](https://github.com/yjs/yjs/blob/fe36ffd122a6f2384293098afd52d2c0025fce2a/src/utils/Doc.js#L53C6-L53C6) is the base `Map` of [shared types](https://github.com/yjs/yjs#shared-types).

MrOrz commented on test/package.json:
Question: what does this package.json do?

nonumpa replied:
It should have a package.json to [define the workspace](https://docs.npmjs.com/cli/v9/using-npm/workspaces?v=true#defining-workspaces).

nonumpa committed:
Update README.md

Co-authored-by: Johnson Liang <johnsonliang7@gmail.com>

nonumpa committed:
Assert error log to be called and ignore optional chaining operator

nonumpa committed:
Test snapshot extension

nonumpa committed:
Remove unused argument in pretest command

nonumpa committed:
Fix test open-handle and jest warning

- Fix hocuspocus open handle by destroying server in each test
- Fix (WARN) Define `ts-jest` config under `globals` is deprecated

MrOrz commented on hocuspocus-extension-elasticsearch/src/__test__/elasticsearch.test.ts:
```diff
@@ -0,0 +1,97 @@
+  };
+  afterEach(async () => {
+    // wait for server async function such as onStoreDocument finished
+    await delayForMs(1000);
```
If we need to wait for a callback in tests, we can consider using done callbacks in individual tests https://jestjs.io/docs/asynchronous#callbacks

MrOrz replied:
Update: `destroy()` may invoke `onStoreDocument()` as a cleanup. `destroy()` does not resolve after all these cleanups, thus we need to wait.

MrOrz commented on src/__test__/snapshot.test.ts:
```diff
@@ -0,0 +1,59 @@
+    });
+
+    // wait for provider to connect and close
+    await delayForMs(1000);
```
Use promise that resolves on `onSynced` to remove this wait

nonumpa committed:
Fix test, remove wait 1s after provider.disconnect()

- Remove wait 1s statement by using `syncedNewHocuspocusProvider` which resolve and return provider after onSynced
- Update 'log error' test to async
- Update comments
